File: shared/src/main/java/ru/privatenull/pnauth/kernel/event/EventDispatchScope.java
package ru.privatenull.pnauth.kernel.event;
final class EventDispatchScope {
    private static final ThreadLocal<ListenerOptions> CURRENT = new ThreadLocal<>();
    static ListenerOptions current() { return CURRENT.get(); }
    static void run(ListenerOptions options, Runnable action) {
        ListenerOptions previous = CURRENT.get(); CURRENT.set(options);
        try { action.run(); } finally { if (previous == null) CURRENT.remove(); else CURRENT.set(previous); }
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/kernel/event/EventDispatchScope.kt
package ru.privatenull.pnauth.kernel.event

internal object EventDispatchScope {
    private val CURRENT = ThreadLocal<ListenerOptions>()

    @JvmStatic
    fun current(): ListenerOptions? = CURRENT.get()

    @JvmStatic
    fun run(options: ListenerOptions, action: Runnable) {
        val previous = CURRENT.get()
        CURRENT.set(options)
        try {
            action.run()
        } finally {
            if (previous == null) CURRENT.remove() else CURRENT.set(previous)
        }
    }
}
